src/dataframe/dataframe.jl

```diff
@@ -1153,13 +1166,49 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame; cols::Symbol=:sete
     try
         for (j, n) in enumerate(_names(df1))
             current_col += 1
-            append!(df1[!, j], df2[!, n])
+            if n in _names(df2)
+                df1_c = df1[!, j]
+                df2_c = df2[!, n]
+                S = eltype(df2_c)
+                T = eltype(df1_c)
+                V = promote_type(S, T)
+                if S <: T || V <: T || !promote
+                    append!(df1_c, df2_c)
+                else
+                    newcol = Tables.allocatecolumn(V, targetrows)
+                    copyto!(newcol, 1, df1_c, 1, nrows)
+                    copyto!(newcol, nrows+1, df2_c, 1, targetrows - nrows)
+                    _columns(df1)[j] = newcol
+                end
+            else
+                # we can safely do it - if we got here this means that this is allowed
+                if Missing <: eltype(df1[!, j]) || !promote
+                    append!(df1[!, j], df2[!, n])
+                else
+                    newcol = Tables.allocatecolumn(Union{Missing,eltype(df1[!, j])} , targetrows)
+                    copyto!(newcol, 1, df1[!, j], 1, nrows)
+                    for i in nrows+1:targetrows
+                        @inbounds newcol[i] = missing
+                    end
+                    _columns(df1)[j] = newcol
+                end
+            end
         end
         current_col = 0
         for col in _columns(df1)
             current_col += 1
             @assert length(col) == targetrows
         end
+        if cols == :union
+            for n in setdiff(_names(df2), _names(df1))
+                newcol = Tables.allocatecolumn(Union{Missing,eltype(df2[!, n])} , targetrows)
+                for i in 1:nrows
+                    @inbounds newcol[i] = missing
+                end
+                copyto!(newcol, nrows+1, df2[!, n], 1, targetrows - nrows)
+                df1[!, n] = newcol
```

Comment:
Why not `_columns(df1)[j] = newcol` here? Or rather, why not `df1[!, n] = newcol` above since it's more standard?

Reply:
This is tricky. Here I use `df1[!, n] = newcol` as it is more standard :smile:, but above `df1[!, n] = newcol` will error when we assign the first column, because it will not have the length equal to number of rows before the assignment (i.e. we are making a data frame taller earlier, so we have to do a workaround).

Reply:
Maybe add a comment when using `_columns` then?

Reply:
added